sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -198,21 +205,39 @@ class Analyzer(override val catalogManager: CatalogManager)
   }
 
   def executeAndCheck(plan: LogicalPlan, tracker: QueryPlanningTracker): LogicalPlan = {
-    if (plan.analyzed) return plan
     AnalysisHelper.markInAnalyzer {
       val analyzed = executeAndTrack(plan, tracker)
       try {
         checkAnalysis(analyzed)
         analyzed
       } catch {
         case e: AnalysisException =>
-          val ae = e.copy(plan = Option(analyzed))
+          val ae = e.copy(plan = Option(analyzed),
+            message = e.message + extraHintForAnsiTypeCoercion(plan))
           ae.setStackTrace(e.getStackTrace)
           throw ae
       }
     }
   }
 
+  private def extraHintForAnsiTypeCoercion(plan: LogicalPlan): String = {
+    if (!conf.ansiEnabled) {
+      ""
+    } else {
+      val nonAnsiPlan = AnalysisContext.withDefaultTypeCoercionAnalysisContext {
+        executeSameContext(plan)
+      }
+      try {
+        checkAnalysis(nonAnsiPlan)
+        "\nTo fix the error, you might need to add explicit type casts.\n" +
```

Comment:
I'm a bit worried about the accuracy here. Re-running the entire analyzer includes more stuff, not just type coercion. Can we be more surgical and only rerun type coercion rules in `CheckAnalysis` when we hit input type mismatch error?

Reply:
Another point is, the analyzer has "side effects", as it may send RPC requests to the remote catalog. I think it's better to not run the entire analyzer again, even if the query fails.

Reply:
> Can we be more surgical and only rerun type coercion rules in CheckAnalysis when we hit input type mismatch error?

We need to rerun some of the rules since they were skipped because the children weren't resolved.
If we have to do it, we should split the case match of `checkAnalysis`  into two parts

Reply:
IIUC the analysis is bottom-up, and `CheckAnalysis` should find the bottom-most expression whose children are all resolved and input type mismatches?

Reply:
Take `ResolveAliases` as an example, the order of it is in front of Type Coercion rules, but it won't happen in the first run since the children is not resolved. After apply the Type Coercion rules, we still have to run the other rules again:
![image](https://user-images.githubusercontent.com/1097932/143989969-cd2af234-a071-4f42-a3f1-bc2c25b55f80.png)